### ISSUE TITLE
Getting exception: missing 2 required positional arguments: 'rpc_controller' and 'request'

### ISSUE THREAD

**Issue by Hakky54** [josephyin/airflow_grpc#1]
Hi thank you creating this library and sharing it. In our project we have applied it but are getting an exception. We have tried different kind of configurations and we are not able to resolve it.

```
[2020-12-09 07:12:36,389] {grpc_operator.py:79} INFO - Calling gRPC service
[2020-12-09 07:12:36,389] {grpc_operator.py:80} INFO - context: {'conf': <airflow.configuration.AirflowConfigParser object at 0x7f8b1336b7f0>, 'dag': <DAG: test>, 'ds': '2020-12-09', 'next_ds': '2020-12-09', 'next_ds_nodash': '20201209', 'prev_ds': '2020-12-09', 'prev_ds_nodash': '20201209', 'ds_nodash': '20201209', 'ts': '2020-12-09T07:12:01.125982+00:00', 'ts_nodash': '20201209T071201', 'ts_nodash_with_tz': '20201209T071201.125982+0000', 'yesterday_ds': '2020-12-08', 'yesterday_ds_nodash': '20201208', 'tomorrow_ds': '2020-12-10', 'tomorrow_ds_nodash': '20201210', 'END_DATE': '2020-12-09', 'end_date': '2020-12-09', 'dag_run': <DagRun test @ 2020-12-09 07:12:01.125982+00:00: manual__2020-12-09T07:12:01.125982+00:00, externally triggered: True>, 'run_id': 'manual__2020-12-09T07:12:01.125982+00:00', 'execution_date': <Pendulum [2020-12-09T07:12:01.125982+00:00]>, 'prev_execution_date': <Pendulum [2020-12-09T07:12:01.125982+00:00]>, 'prev_execution_date_success': <Proxy at 0x7f8aee704588 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7f8b09028620>>, 'prev_start_date_success': <Proxy at 0x7f8b0901d7c8 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7f8b08fb2598>>, 'next_execution_date': <Pendulum [2020-12-09T07:12:01.125982+00:00]>, 'latest_date': '2020-12-09', 'macros': <module 'airflow.macros' from '/home/airflow/.local/lib/python3.6/site-packages/airflow/macros/__init__.py'>, 'params': {}, 'tables': None, 'task': <Task(GrpcOperator): task_id>, 'task_instance': <TaskInstance: test.task_id 2020-12-09T07:12:01.125982+00:00 [running]>, 'ti': <TaskInstance: test.task_id 2020-12-09T07:12:01.125982+00:00 [running]>, 'task_instance_key_str': 'test__task_id__20201209', 'test_mode': False, 'var': {'value': None, 'json': None}, 'inlets': [], 'outlets': []}
[2020-12-09 07:12:36,391] {taskinstance.py:1150} ERROR - <lambda>() missing 2 required positional arguments: 'rpc_controller' and 'request'
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 984, in _run_raw_task
    result = task_copy.execute(context=context)
```

The above happens at the statement `response = rpc_func(**data)` within the hook and at the following statement `self._handle_response(response)` within the operator

Our airflow scripts is:

```
from __future__ import print_function

import datetime

from airflow.models import DAG
from airflow.utils.dates import days_ago
from airflow.operators.dummy_operator import DummyOperator
from google.protobuf.any_pb2 import Any
from google.protobuf.empty_pb2 import _EMPTY

from mycompany import Some_Stub
from mycompany_plugin.operators.grpc_operator import GrpcOperator

args = {
    'owner': 'Airflow',
    'start_date': days_ago(2),
}

dag = DAG(
    dag_id='test',
    default_args=args,
    # schedule_interval='0 7 * * 0', #every Sunday at 7 AM
    schedule_interval='0/30 * * * *',
    tags=['example']
)

def callback(response: Any, **context):
    return response

def empty(args):
    pass

run_this = GrpcOperator(
    task_id='task_id',
    dag=dag,
    grpc_conn_id='grpc_connection_id_on_admin_connections',
    stub_class=Some_Stub,
    call_func='listNames',
    data={},
    response_callback=callback,
    request_data_func=empty
)

with dag:
    dummy = DummyOperator(task_id='kick_off_dag')
    dummy >> run_this
```

Any idea what the cause is and how to resolve it?

**Comment by josephyin** [comment]
pls get the newest version.I fix some issue.And here is some exsample for use GrpcOperator:
```
GrpcOperator(task_id='task_id',
                      dag=dag,
                      grpc_conn_id='your_grpc_connect_address',
                      stub_class=YOUR_GRPC_STUB,
                      call_func='your_grpc_function',
                      request_data_func=YOUR_GRPC_MESSAGE_FOR_REQUEST,
                      response_callback=RESPOSNE_METHOD,
                      xcom_task_id='XCOM_TASK_ID',
                      data={'request':YOUR_REQUEST_DATA_DICT}
```
I think the new version will fix your issue.

**Comment by Hakky54** [comment]
Thank you for replying this quickly. I have adjusted it and now I am getting the following error:
```
[2020-12-09 09:39:42,163] {grpc_operator.py:80} INFO - context: {'conf': <airflow.configuration.AirflowConfigParser object at 0x7f2aaf0be7f0>, 'dag': <DAG: test>, 'ds': '2020-12-09', 'next_ds': '2020-12-09', 'next_ds_nodash': '20201209', 'prev_ds': '2020-12-09', 'prev_ds_nodash': '20201209', 'ds_nodash': '20201209', 'ts': '2020-12-09T09:39:17.175923+00:00', 'ts_nodash': '20201209T093917', 'ts_nodash_with_tz': '20201209T093917.175923+0000', 'yesterday_ds': '2020-12-08', 'yesterday_ds_nodash': '20201208', 'tomorrow_ds': '2020-12-10', 'tomorrow_ds_nodash': '20201210', 'END_DATE': '2020-12-09', 'end_date': '2020-12-09', 'dag_run': <DagRun test @ 2020-12-09 09:39:17.175923+00:00: manual__2020-12-09T09:39:17.175923+00:00, externally triggered: True>, 'run_id': 'manual__2020-12-09T09:39:17.175923+00:00', 'execution_date': <Pendulum [2020-12-09T09:39:17.175923+00:00]>, 'prev_execution_date': <Pendulum [2020-12-09T09:39:17.175923+00:00]>, 'prev_execution_date_success': <Proxy at 0x7f2a8a45f0c8 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7f2aa4d4ed08>>, 'prev_start_date_success': <Proxy at 0x7f2aa4d6fa48 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7f2aa4d78730>>, 'next_execution_date': <Pendulum [2020-12-09T09:39:17.175923+00:00]>, 'latest_date': '2020-12-09', 'macros': <module 'airflow.macros' from '/home/airflow/.local/lib/python3.6/site-packages/airflow/macros/__init__.py'>, 'params': {}, 'tables': None, 'task': <Task(GrpcOperator): task_id>, 'task_instance': <TaskInstance: test.task_id 2020-12-09T09:39:17.175923+00:00 [running]>, 'ti': <TaskInstance: test.task_id 2020-12-09T09:39:17.175923+00:00 [running]>, 'task_instance_key_str': 'test__task_id__20201209', 'test_mode': False, 'var': {'value': None, 'json': None}, 'inlets': [], 'outlets': []}
[2020-12-09 09:39:42,173] {taskinstance.py:1150} ERROR - 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 984, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/opt/airflow/plugins/mycompany_plugin/operators/grpc_operator.py", line 83, in execute
    self.data.update(task_data)
TypeError: 'NoneType' object is not iterable
[2020-12-09 09:39:42,180] {taskinstance.py:1194} INFO - Marking task as FAILED. dag_id=test, task_id=task_id, execution_date=20201209T093917, start_date=20201209T093941, end_date=20201209T093942
[2020-12-09 09:39:46,980] {local_task_job.py:102} INFO - Task exited with return code 1
```

**Comment by josephyin** [comment]
if your xcom is empty. you can just use `data` parameter. 
if your grpc client have no request parameter. `data` and `xcom_task_id` parameter don't need

**Comment by josephyin** [comment]
> pls get the newest version.I fix some issue.And here is some exsample for use GrpcOperator:
> 
> ```
> GrpcOperator(task_id='task_id',
>                       dag=dag,
>                       grpc_conn_id='your_grpc_connect_address',
>                       stub_class=YOUR_GRPC_STUB,
>                       call_func='your_grpc_function',
>                       request_data_func=YOUR_GRPC_MESSAGE_FOR_REQUEST,
>                       response_callback=RESPOSNE_METHOD,
>                       xcom_task_id='XCOM_TASK_ID',
>                       data={'request':YOUR_REQUEST_DATA_DICT}
> ```
> 
> I think the new version will fix your issue.

fix issue. `data={'request':YOUR_REQUEST_DATA_DICT}` just use `data=YOUR_REQUEST_DATA_DICT`

**Comment by Hakky54** [comment]
Our grpc service function does not require parameters. I have removed `data` and `xcom_task_id` but I am now getting my first error:
```
[2020-12-09 10:14:08,258] {grpc_operator.py:80} INFO - context: {'conf': <airflow.configuration.AirflowConfigParser object at 0x7fa1e1da77f0>, 'dag': <DAG: test>, 'ds': '2020-12-09', 'next_ds': '2020-12-09', 'next_ds_nodash': '20201209', 'prev_ds': '2020-12-09', 'prev_ds_nodash': '20201209', 'ds_nodash': '20201209', 'ts': '2020-12-09T10:13:30.701190+00:00', 'ts_nodash': '20201209T101330', 'ts_nodash_with_tz': '20201209T101330.701190+0000', 'yesterday_ds': '2020-12-08', 'yesterday_ds_nodash': '20201208', 'tomorrow_ds': '2020-12-10', 'tomorrow_ds_nodash': '20201210', 'END_DATE': '2020-12-09', 'end_date': '2020-12-09', 'dag_run': <DagRun test @ 2020-12-09 10:13:30.701190+00:00: manual__2020-12-09T10:13:30.701190+00:00, externally triggered: True>, 'run_id': 'manual__2020-12-09T10:13:30.701190+00:00', 'execution_date': <Pendulum [2020-12-09T10:13:30.701190+00:00]>, 'prev_execution_date': <Pendulum [2020-12-09T10:13:30.701190+00:00]>, 'prev_execution_date_success': <Proxy at 0x7fa1d7a385c8 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7fa1d7a33d08>>, 'prev_start_date_success': <Proxy at 0x7fa1d7a5a588 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7fa1d7a5e730>>, 'next_execution_date': <Pendulum [2020-12-09T10:13:30.701190+00:00]>, 'latest_date': '2020-12-09', 'macros': <module 'airflow.macros' from '/home/airflow/.local/lib/python3.6/site-packages/airflow/macros/__init__.py'>, 'params': {}, 'tables': None, 'task': <Task(GrpcOperator): task_id>, 'task_instance': <TaskInstance: test.task_id 2020-12-09T10:13:30.701190+00:00 [running]>, 'ti': <TaskInstance: test.task_id 2020-12-09T10:13:30.701190+00:00 [running]>, 'task_instance_key_str': 'test__task_id__20201209', 'test_mode': False, 'var': {'value': None, 'json': None}, 'inlets': [], 'outlets': []}
[2020-12-09 10:14:08,262] {taskinstance.py:1150} ERROR - <lambda>() missing 2 required positional arguments: 'rpc_controller' and 'request'
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 984, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/opt/airflow/plugins/mycompany_plugin/operators/grpc_operator.py", line 94, in execute
    for response in responses:
  File "/opt/airflow/plugins/mycompany_plugin/hooks/grpc_hook.py", line 57, in run
    response = rpc_func(**data)
TypeError: <lambda>() missing 2 required positional arguments: 'rpc_controller' and 'request'
[2020-12-09 10:14:08,275] {taskinstance.py:1194} INFO - Marking task as FAILED. dag_id=test, task_id=task_id, execution_date=20201209T101330, start_date=20201209T101407, end_date=20201209T101408
[2020-12-09 10:14:12,951] {local_task_job.py:102} INFO - Task exited with return code 1
```

my grpc service descriptor looks like this:

```
  _descriptor.MethodDescriptor(
    name='listNames',
    full_name='mycompany.Some.listNames',
    index=7,
    containing_service=None,
    input_type=google_dot_protobuf_dot_empty__pb2._EMPTY,
    output_type=_LISTNAMESRESPONSE,
    serialized_options=None,
    create_key=_descriptor._internal_create_key,
  ),
```

It looks like it only takes `google_dot_protobuf_dot_empty__pb2._EMPTY` as input parameter, but I am not quite sure if that really counts as an input

By the way you mentioned the following:
>pls get the newest version.I fix some issue.And here is some exsample for use GrpcOperator:

I see in the git commits only changes for the readme but not changes on the actual operator or hook

**Comment by josephyin** [comment]
Oh. I think find some issue. I will fix and test it later.
Can you update your grpc protobuf file to add some parameter and try it by some parameter again?
The readme is not the newest. So I updated it.

**Comment by Hakky54** [comment]
I have retried another function which requires a parameter, but than I am getting the following exception:

```
[2020-12-09 11:00:31,651] {grpc_operator.py:79} INFO - Calling gRPC service
[2020-12-09 11:00:31,651] {grpc_operator.py:80} INFO - context: {'conf': <airflow.configuration.AirflowConfigParser object at 0x7f8f58fc8780>, 'dag': <DAG: test>, 'ds': '2020-12-09', 'next_ds': '2020-12-09', 'next_ds_nodash': '20201209', 'prev_ds': '2020-12-09', 'prev_ds_nodash': '20201209', 'ds_nodash': '20201209', 'ts': '2020-12-09T10:59:55.460604+00:00', 'ts_nodash': '20201209T105955', 'ts_nodash_with_tz': '20201209T105955.460604+0000', 'yesterday_ds': '2020-12-08', 'yesterday_ds_nodash': '20201208', 'tomorrow_ds': '2020-12-10', 'tomorrow_ds_nodash': '20201210', 'END_DATE': '2020-12-09', 'end_date': '2020-12-09', 'dag_run': <DagRun test @ 2020-12-09 10:59:55.460604+00:00: manual__2020-12-09T10:59:55.460604+00:00, externally triggered: True>, 'run_id': 'manual__2020-12-09T10:59:55.460604+00:00', 'execution_date': <Pendulum [2020-12-09T10:59:55.460604+00:00]>, 'prev_execution_date': <Pendulum [2020-12-09T10:59:55.460604+00:00]>, 'prev_execution_date_success': <Proxy at 0x7f8f4ec567c8 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7f8f4ec54d08>>, 'prev_start_date_success': <Proxy at 0x7f8f4ec7b188 with factory <function TaskInstance.get_template_context.<locals>.<lambda> at 0x7f8f4ec7e730>>, 'next_execution_date': <Pendulum [2020-12-09T10:59:55.460604+00:00]>, 'latest_date': '2020-12-09', 'macros': <module 'airflow.macros' from '/home/airflow/.local/lib/python3.6/site-packages/airflow/macros/__init__.py'>, 'params': {}, 'tables': None, 'task': <Task(GrpcOperator): task_id>, 'task_instance': <TaskInstance: test.task_id 2020-12-09T10:59:55.460604+00:00 [running]>, 'ti': <TaskInstance: test.task_id 2020-12-09T10:59:55.460604+00:00 [running]>, 'task_instance_key_str': 'test__task_id__20201209', 'test_mode': False, 'var': {'value': None, 'json': None}, 'inlets': [], 'outlets': []}
[2020-12-09 11:00:31,655] {taskinstance.py:1150} ERROR - <lambda>() missing 1 required positional argument: 'rpc_controller'
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 984, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/opt/airflow/plugins/mycompany_plugin/operators/grpc_operator.py", line 94, in execute
    for response in responses:
  File "/opt/airflow/plugins/mycompany_plugin/hooks/grpc_hook.py", line 57, in run
    response = rpc_func(**data)
TypeError: <lambda>() missing 1 required positional argument: 'rpc_controller'
[2020-12-09 11:00:31,671] {taskinstance.py:1194} INFO - Marking task as FAILED. dag_id=test, task_id=task_id, execution_date=20201209T105955, start_date=20201209T110031, end_date=20201209T110031
```

**Comment by josephyin** [comment]
Which GRPC version you use?
And only can use GRPC client with Python. Are you sure use the python client?

**Comment by Hakky54** [comment]
@josephyin sorry for the late response. We have found the root cause. We used a gradle plugin to generate python grpc code from our proto files. This plugin was unfortunately not generating all the required files. We have solved this by manually generating our grpc python files with the `protoc` command. So your code snippet works fine. I think we can close this issue. Thank you again for solving the other issues in the beginning, that definitely helped as moving forward!
